Implement Reporter#start and Reporter#stop for alternative profiling workflow

NickLaMuro opened the pull request:
## Purpose

When the code block for profiling is not available to the caller of the Profiling code, this API interface allows for starting and stopping the profiler separately, similar to how [`stackprof`](https://github.com/tmm1/stackprof)'s API works.  This maintains the current API and functions equivalently, but allows for a more flexible API for the user.
## Notes
- 0606b17 is both implementing the `.start`/`.stop` methods, and refactoring `.run` to use them.  I originally split this work out into two commits (adding the two methods, then refactoring), but it leads to some extra commit noise and duplicate code that just didn't seem worth it.  If preferred, I can rebase to go back to that.
- I considered adding a `::start` and `::stop` method, but it would have required saving off a class instance variable (I think) to store the current instance of the `Reporter` that was run so `::stop` has something to run against.  I would have preferred to have this interface, but I chose not to include it currently since it would have added extra noise to the PR.  I can add to this one or open up a follow up PR to implement that.
- Possibly an `ensure` should now be included in the `.run` method since we are using the `.start`/`.stop` methods under the hood, and it seems like the source for `ObjectSpace.trace_object_allocations` includes one as well when in the source when calling out to `.trace_object_allocations_start` and `.trace_object_allocations_stop` itself.  I didn't implement this in this PR because:
  1.  My `C` chops are pretty, much non-existent and my understanding of the Ruby source codebase is just as bad
  2.  Still unsure if the `ensure` would be necessary

dgynn commented:
This looks pretty good to me. I haven't come across a situation which would have required this API but it seems like this would work great.

Some thoughts...
* `stackprof` supports starting/stopping the profiling multiple times and then it aggregates the results. So it makes sense that `.stop` wouldn't return the results. `ruby-prof` on the other hand does return results from `.stop` (although it also supports `.suspend`/`.resume`). Should `.stop` return the results so that there is no implication that the report can be stopped/started multiple times?
* A `::start` and `::stop` interface with a class instance variable for the current report makes sense to me. This would also be simplified if `.stop` returned the results.
* Exception handling of the block.call would get more complicated to make sure `trace_object_allocations_stop` gets called and only called once. Instead of `ensure` would it be better to rescue exceptions from `block.call`, stop the tracing, and re-raise the exception? I do see that we currently are **not** handling re-enabling GC if there is an exception in the block. That would be good to add there as well.

@SamSaffron what are your thoughts on adding this API?

SamSaffron commented:
I am fine with adding this API, @NickLaMuro can you add some information about usage to the readme. 

Once @dgynn's concerns/comments are addressed I am fine to have this merged in. 

I like that this PR keeps it all tidy and that we don't need to worry about maintaining two competing codepaths.

NickLaMuro commented:
Cool thanks.  I will look at getting those changes implemented later today then.

NickLaMuro commented:
Responding to some of @dgynn 's comments:

> Should `.stop` return the results so that there is no implication that the report can be stopped/started multiple times?

I could see the multiple `.start`/`.stop` thing for a single report being useful, but I think that might outside the scope of this PR.  I will look into doing that in a separate PR after this one.

> A `::start` and `::stop` interface with a class instance variable for the current report makes sense to me. This would also be simplified if `.stop` returned the results.

Okay, I will the `::start` and `::stop` methods then 😄 .  Regarding returning the report from `.stop`, should I leave that the same for now, or would you prefer that I change that behavior? 

> Exception handling of the `block.call` would get more complicated to make sure `trace_object_allocations_stop` gets called and only called once. Instead of ensure would it be better to rescue exceptions from `block.call`, stop the tracing, and re-raise the exception?

Yeah, I think that is a better way of handling it, and probably less messy.  Will do.

dgynn commented:
Yes, let's change so `.stop` (and `::stop`) returns the results. That method creates the `Results` object so requiring a separate call to get the results seems strange. And I'm not suggesting to add support for multiple start/stop. I want to avoid having the API give the appearance that it is even an option.

SamSaffron commented:
@dgynn @NickLaMuro  I want to close this off, how can we reach closure here? :)

NickLaMuro commented:
@SamSaffron Sorry, this fell of my immediate TODO (though, has been in the back of my mind).  If I don't get back to you in the next day or so with the updates that @dgynn suggested, feel free to close.

Or if for some reason someone wants to finish this up themselves, I won't be offended.

SamSaffron commented:
@NickLaMuro next week is fine 👍

NickLaMuro commented:
@SamSaffron @dgynn Again, sorry for the delay, but I have spent some time and I think I have addressed most of the TODO's on this PR.  Those being mostly around the `::start`/`::stop` methods being added to the API and adding a quick entry into the `README`.

Few things of note regarding what I **didn't** do:

* Didn't touch the `ensure`/exception handling in the `block.call` of `.run` since I was not sure of what we wanted to include in the cleanup.  Figured it should basically be an abbreviated portion of the `stop` method where we are only cleaning up `GC` and `ObjectSpace`, but wasn't sure if that was the right line of thinking with that.
  
* I added some safeguards around the `::start` and `::stop` methods, but wasn't sure if I should make use of those in the `::report` methods as well, preventing someone from doing something like:
  
  ```ruby
  MemoryProfile.report do
    MemoryProfile.start
    # code here
  end
  ```

Curious about your thoughts on that and if they should be done in this PR or others.  Will try and be a little more prompt in responding to review feedback going forward as well ;)

SamSaffron commented:
Maybe we can reuse all our testing as opposed to doubling up ? 

Run tests once using start/stop semantics and a second time using block form?

NickLaMuro commented:
@SamSaffron yeah, I think I have an idea on how we could do that.  I will do that in another commit, and if you don't like what I come up with, we can scrap it.

SamSaffron commented:
definitely better, but I wonder if this can be simply done by having a base class? 

You inherit off the base class with all the tests and implement the logic for run_report_types_for_options in the concrete implementations

NickLaMuro commented:
Yup, never thought of that, but probably would be much cleaner than `.send`.  Will look into trying that out a bit later.

NickLaMuro commented:
@SamSaffron Dropped my previous attempt and replaced them with the most recent three commits which used the base class approach you suggested (I just kept `TestReporter` as the base class in this case for simplicity).  Definitely looks much better and cleaner in the overall diff.

Let me know if you would like me to work some "rebase magic" further to remove the copy-pasta versions of the tests that exists in prior commits of this PR.  Keep the "journey" commits part of it for now, but could see it as adding some necessary noise to the commit history as well.  Up to you.

SamSaffron commented:
Nice, thanks so much! will wait for @dgynn  to have a look and then do a version bump 

merging now 🎊

NickLaMuro commented:
Thanks!  Happy to help.

dgynn commented:
@NickLaMuro This looks really great. Thanks for doing the work to sort out the API changes and get the tests working using shared tests.

When I was looking at the tests, I also wanted to try them with Ruby 2.4.0. The good news is everything still worked. The better news is that Ruby 2.4.0 had even better results. The bad news was that meant one of tests failed in 2.4.0 because fewer objects were allocated.

To fix that, I simplified the way `retained` was handled in the tests and that made it easier to clean up the shared report block creation. The PR is #43. I also merged your 2 test classes into one because it was running the same tests multiple times due to inheritance.

@SamSaffron I think this is good for a version bump and we should be testing 2.4.0 now. Thanks!

NickLaMuro commented:
@dgynn Sorry, this is a bit delayed, but by the time I got this when I first saw it, #43 was already merged, and the changes looked solid, so I didn't have much to add.

Regarding the following though:

> I also merged your 2 test classes into one because it was running the same tests multiple times due to inheritance.

This was actually intentional (though, I am not to terribly tied to/proud of the code by any means), and I was mostly trying to have two separate tests in which one tested the public API, and the other was testing the private one (notice the slight difference in the `create_report` code between the two).  Since the public one (`::start`/`::stop`) is in charge of maintaining an instance of the `Reporter` as a class variable, I figured it made sense to test twice as a way of being able debug issues in the future where that doesn't work properly (hopefully the same tests cases testing the private API would still pass when issues with the `current_reporter` variable arise).

That said, this also might be a bad idea and I am not to concerned with it being removed, just thought I would share the rational as to why it existed in the first place.

Thanks again for looking at this and helping out!

dgynn commented:
@NickLaMuro Good catch! I did miss that distinction in the two `create_report` methods. And testing the class-level API (which I removed) is probably the more important test since that is one that would most likely be used. Would you mind doing a PR to add that test back and making sure it works with 2.4.0?

NickLaMuro commented:
@dgynn Sure thing.  I will take a look into it and ping you when I got something working.

NickLaMuro commented:
@dgynn Here you go!  https://github.com/SamSaffron/memory_profiler/pull/44

Made a stupid mistake last night and forgot to pass in the options hash to the Public version of the tests, causing it to fail (obvious now in hindsight).  Made that realization today after giving up late last night.  Looks like the test pass just fine in ruby 2.4.0.
